src/main/java/net/neoforged/neoforge/transfer/resources/ResourceStack.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.resources;
+
+import java.util.Objects;
+import java.util.function.UnaryOperator;
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.neoforge.fluids.FluidStack;
+
+/**
+ * Represents an immutable {@link IResource} and an amount.
+ * Can be seen as an immutable version of {@link ItemStack} or {@link FluidStack}.
+ */
+public final class ResourceStack<T extends IResource> implements IResourceStack<T> {
+    public static <T extends IResource> ResourceStack<T> of(IResourceStack<T> stack) {
+        return ResourceStack.of(stack.resource(), stack.amount());
+    }
+
+    public static <T extends IResource> ResourceStack<T> of(T resource, int amount) {
+        return new ResourceStack<>(resource, amount);
+    }
+
+    private final T resource;
+    private final int amount;
+
+    private ResourceStack(T resource, int amount) {
+        IResourceStack.validate(resource, amount);
+        this.resource = resource;
+        this.amount = amount;
+    }
+
+    @Override
+    public T resource() {
+        return resource;
+    }
+
+    @Override
+    public int amount() {
+        return amount;
+    }
+
+    /**
+     * @return a copy of this instance with an updated amount.
+     */
+    @Override
+    public ResourceStack<T> withAmount(int newAmount) {
```

Comment:
Given we are immutable, please make this just return itself if the new amount is equal to the current amount, rather than creating a new instance.